File: frontend/src/services/api/suno_service.ts
/**
 * Suno API service for client-side music generation
 */
import Cookies from "js-cookie";

// Update the API base and endpoints to match the correct structure
const SUNO_API_BASE = process.env.NEXT_PUBLIC_SUNO_API_BASE || "https://api.suno.ai";

// Helper function to get the Suno API key
function getSunoApiKey(): string {
  // API key order of precedence: 
  // 1. Environment variable
  // 2. Local storage (if user saved it)
  // 3. Empty string (will trigger appropriate error)
  return process.env.NEXT_PUBLIC_SUNO_API_KEY || 
         localStorage.getItem("sunoApiKey") || 
         "";
}

// Interface for music generation requests
export interface SunoMusicRequest {
  prompt: string;
  make_instrumental?: boolean;
  vocals_only?: boolean;
  custom_mode?: boolean;
  music_style?: string;
  genre?: string;
  vocal_style?: string;
}

// Interface for generation status response
export interface SunoGenerationStatus {
  status: 'pending' | 'processing' | 'completed' | 'failed';
  task_id: string;
  audio_url?: string;
  error_message?: string;
  progress?: number;
  created_at?: string;
  completed_at?: string;
}

/**
 * Create a music generation request with the Suno API
 * @param request The music generation request parameters
 * @returns The generation task information
 */
export async function createSunoGeneration(request: SunoMusicRequest): Promise<{ 
  success: boolean; 
  data?: { task_id: string }; 
  error?: string;
}> {
  try {
    const apiKey = getSunoApiKey();
    
    if (!apiKey) {
      return {
        success: false,
        error: "No Suno API key provided. Please check your environment variables or settings."
      };
    }
    
    // Prepare the request payload based on the Suno API v3 format
    const requestData = {
      prompt: request.prompt,
      ...(request.make_instrumental ? { make_instrumental: true } : {}),
      ...(request.vocals_only ? { vocals_only: true } : {}),
      ...(request.custom_mode && request.music_style ? { music_style: request.music_style } : {}),
      ...(request.custom_mode && request.genre ? { genre: request.genre } : {}),
      ...(request.custom_mode && request.vocal_style ? { vocal_style: request.vocal_style } : {})
    };
    
    // Send the request to Suno API
    const response = await fetch(`${SUNO_API_BASE}/api/v3/generations`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${apiKey}`
      },
      body: JSON.stringify(requestData)
    });
    
    if (!response.ok) {
      const errorText = await response.text();
      console.error("Suno API error:", response.status, errorText);
      return {
        success: false,
        error: `API error (${response.status}): ${errorText}`
      };
    }
    
    const result = await response.json();
    
    return {
      success: true,
      data: {
        task_id: result.id || result.task_id
      }
    };
  } catch (error) {
    console.error("Error creating Suno generation:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error occurred"
    };
  }
}

/**
 * Check the status of a Suno generation task
 * @param taskId The task ID from the creation response
 * @returns The current status information
 */
export async function checkSunoGenerationStatus(taskId: string): Promise<{
  success: boolean;
  data?: SunoGenerationStatus;
  error?: string;
}> {
  try {
    const apiKey = getSunoApiKey();
    
    if (!apiKey) {
      return {
        success: false,
        error: "No Suno API key provided. Please check your environment variables or settings."
      };
    }
    
    const response = await fetch(`${SUNO_API_BASE}/api/v3/generations/${taskId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${apiKey}`
      }
    });
    
    if (!response.ok) {
      const errorText = await response.text();
      console.error("Suno API error:", response.status, errorText);
      return {
        success: false,
        error: `API error (${response.status}): ${errorText}`
      };
    }
    
    const result = await response.json();
    
    // Map Suno API response to our internal status format
    const status = result.status === 'complete' ? 'completed' : 
                  (result.status === 'failed' ? 'failed' : 
                  (result.status === 'in_progress' ? 'processing' : 'pending'));
    
    return {
      success: true,
      data: {
        status: status,
        task_id: taskId,
        audio_url: result.audio_url || result.audio || result.url,
        error_message: result.error || result.error_message,
        progress: result.progress,
        created_at: result.created_at,
        completed_at: result.completed_at
      }
    };
  } catch (error) {
    console.error("Error checking Suno generation status:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error occurred"
    };
  }
}

/**
 * Save Suno API key to local storage (for development/testing only)
 * In production, API keys should be securely stored on the server
 * @param apiKey Suno API key
 */
export function saveSunoApiKey(apiKey: string): void {
  localStorage.setItem("sunoApiKey", apiKey);
}

/**
 * Clear saved Suno API key from local storage
 */
export function clearSunoApiKey(): void {
  localStorage.removeItem("sunoApiKey");
} 

File: backend/mood_based_music/services/suno_service.py
"""
Suno API service module.
Implements comprehensive interactions with the Suno API for music generation.
"""
import os
import json
import logging
import asyncio
import aiohttp
from typing import Dict, Any, List, Optional, Tuple
from django.conf import settings

logger = logging.getLogger(__name__)

class SunoAPIClient:
    """
    Client for interacting with the Suno API.
    Follows the documentation at https://docs.sunoapi.org/
    """
    
    def __init__(self):
        self.api_key = os.getenv('SUNO_API_KEY', settings.SUNO_API_KEY)
        self.api_base = os.getenv('SUNO_API_BASE', settings.SUNO_API_BASE)
        
    async def create_generation(self, 
                                prompt: str, 
                                make_instrumental: bool = False, 
                                vocals_only: bool = False,
                                custom_mode: bool = False,
                                music_style: Optional[str] = None,
                                genre: Optional[str] = None,
                                vocal_style: Optional[str] = None) -> Dict[str, Any]:
        """
        Create a music generation task using Suno API.
        
        Args:
            prompt: The text description prompt for the music
            make_instrumental: Whether to generate instrumental-only track
            vocals_only: Whether to generate vocals-only track
            custom_mode: Whether to use custom mode with more parameters
            music_style: Musical style for custom mode
            genre: Musical genre for custom mode
            vocal_style: Vocal style for custom mode
            
        Returns:
            Dictionary containing the task ID and other information
        """
        create_url = f"{self.api_base}/api/v3/generations"
        
        # Build the generation request payload
        generation_data = {
            "prompt": prompt,
        }
        
        # Add optional parameters
        if make_instrumental:
            generation_data["make_instrumental"] = True
            
        if vocals_only:
            generation_data["vocals_only"] = True
            
        # Add custom mode parameters if applicable
        if custom_mode:
            if music_style:
                generation_data["music_style"] = music_style
            if genre:
                generation_data["genre"] = genre
            if vocal_style:
                generation_data["vocal_style"] = vocal_style
        
        try:
            async with aiohttp.ClientSession() as session:
                # Update headers to use Bearer authentication
                headers = {
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}"
                }
                
                async with session.post(
                    create_url, 
                    headers=headers, 
                    json=generation_data
                ) as response:
                    if response.status != 200 and response.status != 201:
                        error_text = await response.text()
                        error_msg = f"Failed to create generation task: HTTP {response.status} - {error_text}"
                        logger.error(error_msg)
                        raise Exception(error_msg)
                    
                    result = await response.json()
                    
                    # Extract task ID from response
                    task_id = result.get('id')
                    if not task_id:
                        raise Exception("No task ID returned from API")
                    
                    return {
                        "task_id": task_id,
                        "status": "pending",
                        "created_at": result.get('created_at')
                    }
                    
        except Exception as e:
            logger.error(f"Suno API create_generation error: {str(e)}")
            raise
    
    async def get_generation_status(self, task_id: str) -> Dict[str, Any]:
        """
        Check the status of a generation task.
        
        Args:
            task_id: The task ID from the creation response
            
        Returns:
            Dictionary with the current status information
        """
        status_url = f"{self.api_base}/api/v3/generations/{task_id}"
        
        try:
            async with aiohttp.ClientSession() as session:
                # Update headers to use Bearer authentication
                headers = {
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}"
                }
                
                async with session.get(status_url, headers=headers) as response:
                    if response.status != 200:
                        error_text = await response.text()
                        error_msg = f"Failed to check generation status: HTTP {response.status} - {error_text}"
                        logger.error(error_msg)
                        raise Exception(error_msg)
                    
                    result = await response.json()
                    
                    # Map API response to our internal format
                    status = result.get('status', '')
                    if status == 'complete':
                        status = 'completed'
                    elif status == 'in_progress':
                        status = 'processing'
                    
                    return {
                        "task_id": task_id,
                        "status": status,
                        "audio_url": result.get('audio_url') or result.get('url') or result.get('audio'),
                        "error_message": result.get('error'),
                        "progress": result.get('progress'),
                        "created_at": result.get('created_at'),
                        "completed_at": result.get('completed_at')
                    }
                    
        except Exception as e:
            logger.error(f"Suno API get_generation_status error: {str(e)}")
            raise
    
    async def wait_for_completion(self, 
                                 task_id: str, 
                                 timeout_seconds: int = 300, 
                                 check_interval: int = 5) -> Dict[str, Any]:
        """
        Wait for the generation task to complete with timeout.
        
        Args:
            task_id: The task ID to check
            timeout_seconds: Maximum time to wait in seconds
            check_interval: Time between status checks in seconds
            
        Returns:
            Final generation result with audio URL
        """
        max_checks = timeout_seconds // check_interval
        checks = 0
        
        while checks < max_checks:
            result = await self.get_generation_status(task_id)
            
            status = result.get('status')
            if status == 'completed':
                return result
            elif status == 'failed':
                raise Exception(f"Generation task failed: {result.get('error_message', 'Unknown error')}")
            
            # Wait before checking again
            await asyncio.sleep(check_interval)
            checks += 1
            
        raise Exception(f"Generation timed out after {timeout_seconds} seconds")
    
    async def download_audio(self, audio_url: str) -> bytes:
        """
        Download the generated audio file.
        
        Args:
            audio_url: URL of the generated audio
            
        Returns:
            Audio data as bytes
        """
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(audio_url) as response:
                    if response.status != 200:
                        raise Exception(f"Failed to download audio: HTTP {response.status}")
                    
                    return await response.read()
                    
        except Exception as e:
            logger.error(f"Suno API download_audio error: {str(e)}")
            raise
    
    async def generate_music(self, 
                            prompt: str, 
                            wait_for_completion: bool = True,
                            make_instrumental: bool = False,
                            custom_mode: bool = False,
                            **kwargs) -> Tuple[Dict[str, Any], Optional[bytes]]:
        """
        Generate music using the Suno API with one-shot operation.
        
        Args:
            prompt: Text description of the music to generate
            wait_for_completion: Whether to wait for generation to complete
            make_instrumental: Whether to generate instrumental-only track
            custom_mode: Whether to use custom mode
            **kwargs: Additional parameters for custom mode
            
        Returns:
            Tuple of (generation_info, audio_data)
            If wait_for_completion is False, audio_data will be None
        """
        # Create the generation task
        task_result = await self.create_generation(
            prompt=prompt,
            make_instrumental=make_instrumental,
            custom_mode=custom_mode,
            **kwargs
        )
        
        task_id = task_result['task_id']
        
        if not wait_for_completion:
            return task_result, None
        
        # Wait for the task to complete
        completion_result = await self.wait_for_completion(task_id)
        
        # Download the audio
        audio_url = completion_result.get('audio_url')
        if not audio_url:
            raise Exception("No audio URL in completed result")
            
        audio_data = await self.download_audio(audio_url)
        
        return completion_result, audio_data 